### PR TITLE
Retry opening the TSocket in KyuubiSyncThriftClient

### PR DESCRIPTION
**Pull request by mahmoudbahaa**
### _Why are the changes needed?_

the new option: **kyuubi.operation.thrift.client.request.max.attempts** is not respected in createTProtocol in KyuubiSyncThriftClient thus it fails from the first time and doesn't retry even if **kyuubi.operation.thrift.client.request.max.attempts** was geater than 1. so adjusted the code to allow retry for that case. (My scala is kinda rusty so forgive if there is some uncleanness in the code)

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by turboFei** [comment]
FYI:
You can run below command to format the code style.
```
dev/reformat
```

File: kyuubi-server/src/main/scala/org/apache/kyuubi/client/KyuubiSyncThriftClient.scala [hunk 1]
```diff
@@ -106,23 +106,13 @@ class KyuubiSyncThriftClient private (
...
-          throw e
-      }
-    }
+    val (resp, remoteEngineBrokenRetValue) = KyuubiSyncThriftClient.withRetryingRequestNoLock(
+      block,
+      request,
+      maxAttempts,
+      remoteEngineBroken,
+      isConnectionValid)
+    remoteEngineBroken = remoteEngineBrokenRetValue;
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/client/KyuubiSyncThriftClient.scala:115]: The semantics seems a little different with previous. Before we only reset `remoteEngineBroken` if we succeeded to do request, and if failed the `remoteEngineBroken` depend on the probe so we do not change it.

So, we do not pass `remoteEngineBroken` to the new method `withRetryingRequestNoLock`, and change the semantics of the returned value `remoteEngineBrokenRetValue` to `shouldResetRemoteEngineBroken`, then if return true, we change  `remoteEngineBroken` to false, else we do nothing which is same with previous.

Comment (mahmoudbahaa) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/client/KyuubiSyncThriftClient.scala]: @ulysses-you can you please re-check

**Review by ulysses-you — APPROVED** [review]
lgtm, thank you @mahmoudbahaa !

**Review by turboFei — APPROVED** [review]
LGTM

**Comment by turboFei** [comment]
thanks, merged to master

**Pull request closed without merging**